test/functional/lua/iter_spec.lua

```diff
@@ -390,6 +390,37 @@ describe('vim.iter', function()
     end))
   end)
 
+  it('flatten()', function()
+    local t = { { 1, { 2 } }, { { { { 3 } } }, { 4 } }, { 5 } }
+
+    eq({ 1, { 2 }, { { { 3 } } }, { 4 }, 5 }, vim.iter(t):flatten():totable())
+    eq({ 1, 2, { { 3 } }, 4, 5 }, vim.iter(t):flatten(2):totable())
+    eq({ 1, 2, { 3 }, 4, 5 }, vim.iter(t):flatten(3):totable())
+    eq({ 1, 2, 3, 4, 5 }, vim.iter(t):flatten(4):totable())
+
+    local corner = { 1, { 2, 3, nil, 4 }, 5, { 6, a = 7, b = 8 } }
+    eq({ 1, 2, 3, 4, 5, 6 }, vim.iter(corner):flatten():totable())
```

Comment:
If a value looks like a dict-like table, it should be preserved. We should be able to flatten lists of dicts.

    eq({ { a = 1}, { b = 2} }, vim.iter({ { { a = 1}, { b = 2} } }):flatten():totable())